dipy/reconst/weights_method.py

```diff
@@ -0,0 +1,248 @@
+                              m_est="gm", cutoff=3,
+                              outlier_condition_func=simple_cutoff):
+    """ Calculate M-estimator weights for NLLS model.
+
+    Parameters
+    ----------
+    data : array
+        The measured signal.
+    pred_sig : array
+        The predicted signal, given a previous fit.
+        Has the same shape as data.
+    design_matrix : array (g, ...)
+        Design matrix holding the covariants used to solve for the
+        regression coefficients.
+    leverages : array
+        The leverages (diagonal of the 'hat matrix') of the fit.
+    idx : int
+        The current iteration number.
+    total_idx : int
+        The total number of iterations.
+    last_robust : array (int),
+        Array of 1s (inlier) and 0s (outlier).
+        Has the same shape as data.
+    m_est : which M-estimator weighting scheme to use. Currently,
+            'gm' (Geman-McClure) and 'cauchy' are provided.
+    cut_off : float
+        Outliers are defined by residuals > |cut_off * standard_deviation|
+
+    Notes
+    -----
+    Robust weights are calculated specifically for the NLLS problem.
+    On the last iteration, NLLS is performed without outliers.
+
+    """
+    # check if M-estimator is valid (defined in this function)
+    if m_est not in ["gm", "cauchy"]:
+        raise ValueError("unknown M-estimator choice")
+
+    # min 3 iters: (1) NLLS (2) NLLS with M-weights (3) clean NLLS
+    if total_idx < 3:
+        raise ValueError("require 3+ iterations")
+
+    p, N = design_matrix.shape[-1], data.shape[-1]
+    if N <= p:
+        raise ValueError("Fewer data points than parameters.")
+    factor = 1.4826 * np.sqrt(N / (N - p))
+
+    # handle potential zeros
+    pred_sig[pred_sig < MIN_POSITIVE_SIGNAL] = MIN_POSITIVE_SIGNAL
+
+    # calculate quantities needed for C and w
+    log_pred_sig = np.log(pred_sig)
+    residuals = data - pred_sig
+    log_data = np.log(data)
+    log_residuals = log_data - log_pred_sig
+    z = pred_sig * log_residuals
+
+    C = factor * np.median(np.abs(residuals - np.median(residuals)[..., None]),
+                           axis=-1)[..., None]
+    C[C == 0] = np.nanmedian(C)  # C could be 0, if all signals = min_signal
+
+    # NOTE: if more M-estimators are added, please update the docs!
+    if m_est == "gm":
+        w = C**2 / (C**2 + residuals**2)**2
+    if m_est == "cauchy":
+        w = C**2 / (C**2 + residuals**2)
+
+    robust = None
+
+    if idx == total_idx:
+        cond = outlier_condition_func(
+            residuals, log_residuals, pred_sig,
+            design_matrix, leverages, C, cutoff
+        )
+        robust = np.logical_not(cond)
+
+        w[robust == 0] = 0.0
+        w[robust == 1] = 1.0
+
+    w[np.isinf(w)] = 0
+    w[np.isnan(w)] = 0
+
+    return w, robust
+
+
+# define specific M-estimators
+def weights_method_wls_gm(*args):
+    """ return weights_method_wls_m_est(*args, m_est="gm", cutoff=3),
```

Comment:
A few more notes about docstrings applying to the entire PR.
- Remove the whitespace before the beginning of the short description.
- Start the sentences with capitals.
- Use a short description/summary (one line), and then start a new paragraph (blank line between them) with the long description so that the documentation can be rendered correctly.

Thanks.